tests/ClientBuilderTest.php

```diff
@@ -34,127 +35,142 @@ public function testCreate(): void
-    }
-
-    public function testNullTransportIsUsedWhenNoServerIsConfigured(): void
-    {
-        $clientBuilder = new ClientBuilder();
-
-        $transport = $this->getObjectAttribute($clientBuilder->getClient(), 'transport');
+        $options = new Options();
+        $clientBuilder = new ClientBuilder($options);
 
-        $this->assertInstanceOf(NullTransport::class, $transport);
+        $this->assertSame($options, $clientBuilder->getOptions());
     }
 
-    public function testSetUriFactory(): void
+    public function testAddHttpClientPlugin(): void
     {
-        /** @var UriFactory|MockObject $uriFactory */
-        $uriFactory = $this->createMock(UriFactory::class);
+        /** @var Plugin|MockObject $plugin */
+        $plugin = $this->createMock(Plugin::class);
+        $plugin->expects($this->once())
+            ->method('handleRequest')
+            ->willReturnCallback(function (RequestInterface $request, callable $next) {
+                return $next($request);
+            });
 
-        $clientBuilder = ClientBuilder::create(['dsn' => 'http://public:secret@example.com/sentry/1']);
-        $clientBuilder->setUriFactory($uriFactory);
+        $client = ClientBuilder::create(['dsn' => 'http://public:secret@example.com/sentry/1'])
+            ->addHttpClientPlugin($plugin)
+            ->getClient();
 
-        $this->assertAttributeSame($uriFactory, 'uriFactory', $clientBuilder);
+        $client->captureEvent([]);
     }
 
-    public function testSetMessageFactory(): void
+    public function testRemoveHttpClientPlugin(): void
     {
-        /** @var MessageFactory|MockObject $messageFactory */
-        $messageFactory = $this->createMock(MessageFactory::class);
-
-        $clientBuilder = ClientBuilder::create(['dsn' => 'http://public:secret@example.com/sentry/1']);
-        $clientBuilder->setMessageFactory($messageFactory);
-
-        $this->assertAttributeSame($messageFactory, 'messageFactory', $clientBuilder);
-
-        $transport = $this->getObjectAttribute($clientBuilder->getClient(), 'transport');
-
-        $this->assertAttributeSame($messageFactory, 'requestFactory', $transport);
+        /** @var Plugin|MockObject $plugin1 */
+        $plugin1 = $this->getMockBuilder(Plugin::class)
+            ->setMockClassName('PluginMock1')
+            ->getMock();
+
+        $plugin1->expects($this->exactly(2))
+            ->method('handleRequest')
+            ->willReturnCallback(function (RequestInterface $request, callable $next) {
+                return $next($request);
+            });
+
+        /** @var Plugin|MockObject $plugin2 */
+        $plugin2 = $this->createMock(Plugin::class);
+        $plugin2->expects($this->never())
+            ->method('handleRequest');
+
+        $client = ClientBuilder::create(['dsn' => 'http://public:secret@example.com/sentry/1'])
+            ->addHttpClientPlugin($plugin1)
+            ->addHttpClientPlugin($plugin1)
+            ->addHttpClientPlugin($plugin2)
```

Comment:
Could you move this add in the middle of the other two? Just to be sure that we're not popping a generic plugin, but instead removing the requested one.